src/main/java/net/glowstone/net/handler/legacyping/LegacyPingHandler.java

```diff
@@ -0,0 +1,101 @@
+package net.glowstone.net.handler.legacyping;
+
+import java.io.UnsupportedEncodingException;
+import java.net.InetSocketAddress;
+
+import org.bukkit.event.server.ServerListPingEvent;
+
+import com.google.common.base.Charsets;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import net.glowstone.EventFactory;
+import net.glowstone.GlowServer;
+import net.glowstone.net.GlowNetworkServer;
+
+public class LegacyPingHandler extends ChannelInboundHandlerAdapter {
+
+    GlowNetworkServer networkServer;
+
+    public LegacyPingHandler(GlowNetworkServer networkServer) {
+        this.networkServer = networkServer;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext channelHandlerContext, Object object) throws Exception {
+        ByteBuf bytebuf = (ByteBuf) object;
+
+        bytebuf.markReaderIndex();
+        boolean legacyPingProtocol = false;
+
+        try {
+            if (bytebuf.readByte() == (byte) 0xFE) {
+                GlowServer server = networkServer.getServer();
+                int readableBytes = bytebuf.readableBytes();
+
+                InetSocketAddress inetsocketaddress = (InetSocketAddress) channelHandlerContext.channel().remoteAddress();
+
+                ServerListPingEvent legacyPingEvent = new ServerListPingEvent(inetsocketaddress.getAddress(), server.getMotd(), server.getOnlinePlayers().size(), server.getMaxPlayers());
```

Comment:
I am confused on what you are saying I should change :confused: . I understand that it will wait for the main thread (which will introduce minor lag) but that allows for plugins to modify the response. Isn't that what the event is for? If I'm completely misunderstanding you could you restate what you said.